application_sandbox/separate_stencil_usage/main.cpp

```diff
@@ -0,0 +1,649 @@
+  containers::unique_ptr<vulkan::VkImageView> depth_stencil_view_stencil_only_;
+};
+
+class SeparateStencilUsageSample
+    : public sample_application::Sample<SeparateStencilUsageFrameData> {
+ public:
+  SeparateStencilUsageSample(const entry::EntryData* data)
+      : data_(data),
+        Sample<SeparateStencilUsageFrameData>(
+            data->allocator(), data, 1, 512, 1, 1,
+            sample_application::SampleOptions().EnableVulkan11(), {0}, {},
+            {VK_EXT_SEPARATE_STENCIL_USAGE_EXTENSION_NAME}),
+        cube_(data->allocator(), data->logger(), cube_data),
+        floor_(data->allocator(), data->logger(), floor_data),
+        plane_(data->allocator(), data->logger(), plane_data) {}
+
+  virtual void InitializeApplicationData(
+      vulkan::VkCommandBuffer* initialization_buffer,
+      size_t num_swapchain_images) override {
+    cube_.InitializeData(app(), initialization_buffer);
+    floor_.InitializeData(app(), initialization_buffer);
+    plane_.InitializeData(app(), initialization_buffer);
+
+    descriptor_set_layouts_[0] = {
+        0,                                  // binding
+        VK_DESCRIPTOR_TYPE_UNIFORM_BUFFER,  // descriptorType
+        1,                                  // descriptorCount
+        VK_SHADER_STAGE_VERTEX_BIT,         // stageFlags
+        nullptr                             // pImmutableSamplers
+    };
+    descriptor_set_layouts_[1] = {
+        1,                                  // binding
+        VK_DESCRIPTOR_TYPE_UNIFORM_BUFFER,  // descriptorType
+        1,                                  // descriptorCount
+        VK_SHADER_STAGE_VERTEX_BIT,         // stageFlags
+        nullptr                             // pImmutableSamplers
+    };
+
+    pipeline_layout_ = containers::make_unique<vulkan::PipelineLayout>(
+        data_->allocator(),
+        app()->CreatePipelineLayout(
+            {{descriptor_set_layouts_[0], descriptor_set_layouts_[1]}}));
+
+    read_stencil_pipeline_layout_bindings_ = {
+        0,                                    // binding
+        VK_DESCRIPTOR_TYPE_INPUT_ATTACHMENT,  // descriptorType
+        1,                                    // descriptorCount
+        VK_SHADER_STAGE_FRAGMENT_BIT,         // stage
+        nullptr                               // pImmutableSamplers
+    };
+
+    read_stencil_pipeline_layout_ =
+        containers::make_unique<vulkan::PipelineLayout>(
+            data_->allocator(),
+            app()->CreatePipelineLayout(
+                {{read_stencil_pipeline_layout_bindings_}}));
+
+    VkAttachmentReference color_attachment = {
+        0, VK_IMAGE_LAYOUT_COLOR_ATTACHMENT_OPTIMAL};
+    VkAttachmentReference depth_attachment = {
+        1, VK_IMAGE_LAYOUT_DEPTH_STENCIL_ATTACHMENT_OPTIMAL};
+    VkAttachmentReference read_stencil_attachment = {
+        1, VK_IMAGE_LAYOUT_DEPTH_STENCIL_READ_ONLY_OPTIMAL};
+
+    render_pass_ = containers::make_unique<vulkan::VkRenderPass>(
```

Comment:
From what I can tell of your usage: this should be trivial to do with 2 subpasses && a single render-pass, instead of 2 complete renderpasses.

Reply:
Yeah. I had tried that initially but couldn't get it to work. I think that was mainly due to my complete lack of understanding of how pipeline barriers work. After much struggling I finally got it to work with 2 completely separate passes and forgot to revisit this. I was able able to easily migrate to 2 subpasses now that the barriers/layouts/stages work.